Bricks/file/file.h

```diff
@@ -214,79 +214,125 @@ struct FileSystem {
-          // with Bricks' directory mounted from Mac's filesystem.
-          // `entry->d_type` is always zero there, see http://comments.gmane.org/gmane.comp.lib.libcg.devel/4236
-          const ScanDirParameters mask =
-              IsDir(JoinPath(directory, filename)) ? ScanDirParameters::ListDirsOnly : ScanDirParameters::ListFilesOnly;
-          if (static_cast<int>(parameters) & static_cast<int>(mask)) {
-            if (!f(filename)) {
-              return;
+        while (struct dirent* entry = ::readdir(dir)) {
+          const char* const name = entry->d_name;
+          if (ScanDirCanHandleName(name)) {
+            // `IsDir` is proved to be required on Ubuntu running in Parallels on a Mac,
+            // with Bricks' directory mounted from Mac's filesystem.
+            // `entry->d_type` is always zero there, see http://comments.gmane.org/gmane.comp.lib.libcg.devel/4236
+            const std::string& path = JoinPath(directory, name);
+            const bool is_directory = IsDir(path);
+            const ScanDirParameters mask =
+                is_directory ? ScanDirParameters::ListDirsOnly : ScanDirParameters::ListFilesOnly;
+            if (static_cast<int>(parameters) & static_cast<int>(mask)) {
+              if (!item_handler(ScanDirItemInfo(name, path, is_directory))) {
+                return;
+              }
             }
           }
         }
-      }
-    } else {
-      if (errno == ENOENT) {
-        CURRENT_THROW(DirDoesNotExistException(directory));
-      } else if (errno == ENOTDIR) {
-        CURRENT_THROW(PathNotDirException(directory));
       } else {
-        CURRENT_THROW(FileException(directory));  // LCOV_EXCL_LINE
+        if (errno == ENOENT) {
+          CURRENT_THROW(DirDoesNotExistException(directory));
+        } else if (errno == ENOTDIR) {
+          CURRENT_THROW(PathNotDirException(directory));
+        } else {
+          CURRENT_THROW(FileException(directory));  // LCOV_EXCL_LINE
+        }
       }
-    }
 #endif
+    } else {
+      const std::function<bool(const ScanDirItemInfo&)> item_handler_recursive = [&item_handler, parameters](const ScanDirItemInfo& item_info) {
+        const ScanDirParameters mask = item_info.is_directory
+                                         ? ScanDirParameters::ListDirsOnly
+                                         : ScanDirParameters::ListFilesOnly;
+        if (static_cast<int>(parameters) & static_cast<int>(mask)) {
+          if (!item_handler(item_info)) {
+            return false;
+          }
+        }
+        if (item_info.is_directory) {
+          ScanDirUntil<ITEM_HANDLER>(item_info.path, std::forward<ITEM_HANDLER>(item_handler), parameters, ScanDirRecursive::Yes);
+        }
+        return true;
+      };
+      ScanDirUntil(directory,
+                   item_handler_recursive,
+                   ScanDirParameters::ListFilesAndDirs,
+                   ScanDirRecursive::No);
+    }
```

Comment:
This part looks super complex. Why not just remove top-level `if (recursive == ScanDirRecursive::No)` and augment processing loop contents with:
```
if (is_directory && recursive == ScanDirRecursive::Yes) {
  ScanDirUntil(path, item_handler, parameters, recursive);
}
```